File: 4 Estrategias Evolutivas/Códigos/1 (1+1)-ES/ES.py
import numpy as np
from Plot_Contour import plot_contour
from Plot_Surf import plot_surf

# f = lambda x, y: x * np.exp((-x**2) - (y**2))
# xl = np.array([-2, -2])
# xu = np.array([2, 2])

f = lambda x, y: (x-2)**2 + (y-2)**2
xl = np.array([-5, -5])
xu = np.array([5, 5])

G = 100
D = 2

sigma = 0.2

x = xl + (xu - xl) * np.random.rand(D)

for i in range(G):
    r = np.random.normal(0, sigma, D)
    y = x + r

    if f(y[0], y[1]) < f(x[0], x[1]):
        x = y.copy()

    plot_contour(f,x, y, xl, xu)

plot_surf(f, x, xl, xu)
print("Mínimo en x=", x[0], " y=", x[1], " f(x,y)=", f(x[0], x[1]))

File: 4 Estrategias Evolutivas/Códigos/1 (1+1)-ES/ES.m
clear all
close all
clc

f = @(x,y) x.*exp(-x.^2-y.^2);
xl = [-2 -2]';
xu = [2 2]';

% f = @(x,y) (x-2).^2+(y-2).^2;
% xl = [-5 -5]';
% xu = [5 5]';

G = 100;
D = 2;

sigma = 0.2;

x = xl+(xu-xl).*rand(D,1);

for t=1:G
	r = normrnd(0,sigma,[D 1]);
    y = x + r;
    
    if f(y(1),y(2)) < f(x(1),x(2))
        x = y;
    end

    Plot_Contour(f,[x y],xl,xu); % Gráfica
end

figure
Plot_Surf(f,[x y],xl,xu) % Gráfica
disp([' mínimo global en: x=' num2str(x(1)) ', y=' num2str(x(2)) ', f(x,y)=' num2str(f(x(1),x(2)))])
